internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher.go

```diff
@@ -93,52 +115,328 @@ func (c *gitHubWatcherController) Sync(ctx controllerlib.Context) error {
+
+		return conditions, hostWithHttps, nil, nil
+	}
+
+	tlsDialer := tls.Dialer{
+		NetDialer: &net.Dialer{
+			Timeout: 30 * time.Second,
+		},
+		Config: &tls.Config{
+			MinVersion: tls.VersionTLS12,
+			// If certPool is nil then RootCAs will be set to nil and TLS will use the host's root CA set automatically.
+			RootCAs: certPool,
+		},
+	}
+
+	conn, tlsDialErr := tlsDialer.DialContext(ctx, "tcp", hostPort.Endpoint())
+	if tlsDialErr != nil {
+		conditions = append(conditions, &metav1.Condition{
+			Type:    TLSConfigurationValid,
+			Status:  metav1.ConditionFalse,
+			Reason:  upstreamwatchers.ReasonInvalidTLSConfig,
+			Message: fmt.Sprintf("tls configuration is not valid for spec.githubAPI.host (%q): %s", host, tlsDialErr.Error()),
+		})
+		conditions = append(conditions, githubConnectionUnknown)
+		return conditions, hostWithHttps, nil, tlsDialErr
+	}
+
+	conditions = append(conditions, &metav1.Condition{
+		Type:    TLSConfigurationValid,
+		Status:  metav1.ConditionTrue,
+		Reason:  upstreamwatchers.ReasonSuccess,
+		Message: fmt.Sprintf("tls configuration is valid for spec.githubAPI.host (%q)", host),
+	})
+	conditions = append(conditions, &metav1.Condition{
+		Type:    GitHubConnectionValid,
+		Status:  metav1.ConditionTrue,
+		Reason:  upstreamwatchers.ReasonSuccess,
+		Message: fmt.Sprintf("spec.githubAPI.host (%q) is reachable and TLS verification succeeds", host),
+	})
+
+	return conditions, hostWithHttps, c.httpClientBuilder(certPool), conn.Close()
+}
+
+func validateUserAndGroupAttributes(upstream *v1alpha1.GitHubIdentityProvider) (v1alpha1.GitHubGroupNameAttribute, v1alpha1.GitHubUsernameAttribute, error) {
+	var groupNameAttribute v1alpha1.GitHubGroupNameAttribute
+	var usernameAttribute v1alpha1.GitHubUsernameAttribute
+
+	if upstream.Spec.Claims.Groups == nil || upstream.Spec.Claims.Username == nil {
+		return "", "", fmt.Errorf("spec.claims.groups and spec.claims.username are required")
+	} else {
+		groupNameAttribute = *upstream.Spec.Claims.Groups
+		usernameAttribute = *upstream.Spec.Claims.Username
+	}
+
+	// Should not happen due to CRD enum validation
+	switch usernameAttribute {
+	case v1alpha1.GitHubUsernameLoginAndID:
+	case v1alpha1.GitHubUsernameLogin:
+	case v1alpha1.GitHubUsernameID:
+		break
+	default:
+		return "", "", fmt.Errorf("invalid spec.claims.username")
+	}
+
+	// Should not happen due to CRD enum validation
+	switch groupNameAttribute {
+	case v1alpha1.GitHubUseTeamNameForGroupName:
+	case v1alpha1.GitHubUseTeamSlugForGroupName:
+		break
+	default:
+		return "", "", fmt.Errorf("invalid spec.claims.groups")
+	}
+
+	return groupNameAttribute, usernameAttribute, nil
+}
+
+// buildCertPool returns a PEM-encoded CA bundle from the provided spec. If the provided spec is nil, a
+// nil CA bundle will be returned. If the provided spec contains a CA bundle that is not properly
+// encoded, an error will be returned.
+func buildCertPool(spec *v1alpha1.TLSSpec) (*x509.CertPool, []byte, error) {
```

Comment:
Do we have this helper somewhere?

Reply:
Shall we pull this out and make it a sibling helper to the other, perhaps?

Reply:
helper extracted